core/src/main/java/com/amplifyframework/core/DeveloperMenuActivity.java

```diff
@@ -29,27 +27,34 @@
  * This is the activity to display the developer menu.
  */
 public final class DeveloperMenuActivity extends FragmentActivity {
+
+    // Indicates whether this activity has started.
+    private static boolean activityStarted;
```

Comment:
Maintaining state in a static variable should generally be avoided because it can be accessed/modified from anywhere, and thus can lead to bugs and code that's hard to maintain.  Looks like you are referencing this in `Amplify.configure` to see whether DeveloperMenuActivity should be started.  Instead, I'd suggest this:
 - When a shake is detected, stop listening for shakes, and then start DeveloperMenuActivity using the `startActivityForResult()` method (https://developer.android.com/training/basics/intents/result). 
 - Then, implement `onActivityResult()` to know when DeveloperMenuActivity is finished.  When it is finished, start listening for shakes again.

Reply:
For our use case, would this approach be equivalent to starting to listen for shakes again in the DeveloperMenuActivity onStop()?

Reply:
Hmm, how are you referencing the ShakeDetector defined in Amplify.configure from `DeveloperMenuActivity.onStop()`

One concern is that the logic for starting and stopping listening is not encapsulated in one place, and thus harder to maintain/understand - starting is in Amplify.configure, and stopping is in `DeveloperMenuActivity`. 

Another concern is the following scenario:
 - Amplify.configure starts listening for shakes.
 - Customer shakes device, listening stops, DeveloperMenuActivity is launched.
 - Customer receives a push notification, taps it, which launches a new Activity to display that push notification.  `onStop` is called, so listening starts again.
 - Customer taps back, and is now on DeveloperMenuActivity, and listener is still active. 
 - Customer shakes devices again, and now a new DeveloperMenuActivity instance is started on top of the existing one.

Reply:
In DeveloperMenuActivity.onStop(), I call Amplify.enableDeveloperMenu, which handles starting and stopping listening for shake events when the developer menu activity has not started (essentially creating a new ShakeDetector instance).

Reply:
I see - hmm, we really should avoid making `Amplify.enableDeveloperMenu` a public method, as that is now exposed to customers, and I don't think there's a strong customer use case to support that.  If there was a customer use case, this is the kind of thing that would need to go through a design review with all platforms as well.

Reply:
I didn't realize that, thanks for pointing it out. I was trying to avoid repeating the same code in DeveloperMenuActivity. If I move the method to DeveloperMenuActivity, but still keep it public so I can call it in Amplify.configure, would it no longer be exposed to customers?

Reply:
My previous suggestion was to implement `onActivityResult` in the class that starts `DeveloperMenuActivity`, but I just realized that doesn't work because that has to be implemented in an Activity, and `Amplify` is not an `Activity`. 

Still thinking about this...

Reply:
Since I'm starting the developer menu activity from outside of an activity, I have to use FLAG_ACTIVITY_NEW_TASK. The documentation for this flag says "if a task is already running for the activity you are now starting, then a new activity will not be started; instead, the current task will simply be brought to the front of the screen with the state it was last in". Does this mean that in the scenario you described, a new developer menu activity would not be started?

Reply:
> Does this mean that in the scenario you described, a new developer menu activity would not be started?

Yes, that's right.  

If you are okay with not supporting shake to dismiss, using the `FLAG_ACTIVITY_SINGLE_TOP` would probably give you a complete solution.    

If you do want to support shake to dismiss, it is a bit more complex.  I think you could accomplish it though with something like the logic below.  It's not my favorite idea, because it still involves maintaining a static state, but I'm not sure of a different solution - it's definitely a tricky problem.

```
class DevMenuManager implements ShakeDetector.Listener {
    private boolean visible;
 
    private static DevMenuManager sInstance = new DevMenuManager();

    public static sharedInstance() {
         return sInstance;
    }

    public start() {
         // Subscribe to shake events
    }

    public void onShakeDetected() {
         if(visible) {
            Amplify.Hub.sendMessage("HideDeveloperMenu");  // this syntax is probably wrong, but the idea is to broadcast a message on Hub.
         } else {
            Intent mainIntent = new Intent(context, DeveloperMenuActivity.class);
            mainIntent.setAction(Intent.ACTION_MAIN);
            mainIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(mainIntent);
            visible = true;
       }
   }

   public void setHidden() {
       this.visible = false;
   }
}
```

In `Amplify.configure()`:

```
DevMenuManager.sharedInstance().startListening()
```

In `DeveloperMenuActivity.onStop()`:

```
DevMenuManager.sharedInstance.setHidden();
```